File: sort-mapper-example-java/src/main/java/kh/farrukh/sortmapperexamplejava/lesson/LessonServiceImpl.java
package kh.farrukh.sortmapperexamplejava.lesson;

import java.time.LocalDateTime;
import java.util.Optional;
import kh.farrukh.sortmapperexamplejava.common.exception.NotFoundException;
import kh.farrukh.sortmapperexamplejava.course.CourseService;
import kh.farrukh.sortmapperexamplejava.lesson.model.LessonCreateRequestDTO;
import kh.farrukh.sortmapperexamplejava.lesson.model.LessonDetailsResponseDTO;
import kh.farrukh.sortmapperexamplejava.lesson.model.LessonResponseDTO;
import kh.farrukh.sortmapperexamplejava.lesson.model.LessonUpdateRequestDTO;
import kh.farrukh.sortmapperexamplejava.teacher.TeacherService;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class LessonServiceImpl implements LessonService {

  private final LessonRepo lessonRepo;
  private final LessonMapper lessonMapper;
  private final TeacherService teacherService;
  private final CourseService courseService;

  @Override
  public void createLesson(LessonCreateRequestDTO requestDTO) {
    var lesson = lessonMapper.toEntity(requestDTO);
    var course = courseService.findCourse(requestDTO.courseId());
    var teacher = Optional.ofNullable(requestDTO.helperTeacherId())
        .map(teacherService::findTeacher)
        .orElse(null);
    lesson.setCourse(course);
    lesson.setHelperTeacher(teacher);
    lessonRepo.save(lesson);
  }

  @Override
  public LessonDetailsResponseDTO getLesson(Long id) {
    return lessonMapper.toDetailsResponseDTO(findLesson(id));
  }

  @Override
  public Page<LessonResponseDTO> getLessons(Pageable pageable) {
    return lessonRepo
        .findAllByDeletedAtIsNull(pageable)
        .map(lessonMapper::toResponseDTO);
  }

  @Override
  public void updateLesson(Long id, LessonUpdateRequestDTO requestDTO) {
    var lesson = findLesson(id);
    lessonMapper.update(lesson, requestDTO);
    var course = courseService.findCourse(requestDTO.courseId());
    var teacher = Optional.ofNullable(requestDTO.helperTeacherId())
        .map(teacherService::findTeacher)
        .orElse(null);
    lesson.setCourse(course);
    lesson.setHelperTeacher(teacher);
    lessonRepo.save(lesson);
  }

  @Override
  public void deleteLesson(Long id) {
    var lesson = findLesson(id);
    lesson.setDeletedAt(LocalDateTime.now());
    lessonRepo.save(lesson);
  }

  @Override
  public LessonEntity findLesson(Long id) {
    return lessonRepo
        .findByIdAndDeletedAtIsNull(id)
        .orElseThrow(() -> new NotFoundException("Lesson not found with id: " + id));
  }

}

File: sort-mapper-example/src/main/kotlin/kh/farrukh/sortmapperexample/lesson/LessonServiceImpl.kt
package kh.farrukh.sortmapperexample.lesson

import kh.farrukh.sortmapperexample.common.exception.NotFoundException
import kh.farrukh.sortmapperexample.course.CourseService
import kh.farrukh.sortmapperexample.lesson.model.LessonCreateRequestDTO
import kh.farrukh.sortmapperexample.lesson.model.LessonDetailsResponseDTO
import kh.farrukh.sortmapperexample.lesson.model.LessonResponseDTO
import kh.farrukh.sortmapperexample.lesson.model.LessonUpdateRequestDTO
import kh.farrukh.sortmapperexample.teacher.TeacherService
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class LessonServiceImpl(
    private val lessonRepo: LessonRepo,
    private val courseService: CourseService,
    private val teacherService: TeacherService
) : LessonService {

    override fun createLesson(requestDTO: LessonCreateRequestDTO) {
        val course = courseService.findCourse(requestDTO.courseId)
        val helperTeacher = requestDTO.helperTeacherId?.let(teacherService::findTeacher)
        val lesson = requestDTO.toEntity(course, helperTeacher)
        lessonRepo.save(lesson)
    }

    override fun getLessonById(id: Long): LessonDetailsResponseDTO {
        return findLesson(id).toDetailsResponseDTO()
    }

    override fun getAllLessons(pageable: Pageable): Page<LessonResponseDTO> {
        return lessonRepo
            .findAllByDeletedAtIsNull(pageable)
            .map(LessonEntity::toResponseDTO)
    }

    override fun updateLesson(id: Long, requestDTO: LessonUpdateRequestDTO) {
        val lesson = findLesson(id)
        val course = courseService.findCourse(requestDTO.courseId)
        val helperTeacher = requestDTO.helperTeacherId?.let(teacherService::findTeacher)
        lesson.update(requestDTO, course, helperTeacher)
        lessonRepo.save(lesson)
    }

    override fun deleteLesson(id: Long) {
        val lesson = findLesson(id)
        lesson.deletedAt = LocalDateTime.now()
        lessonRepo.save(lesson)
    }

    private fun findLesson(id: Long) = lessonRepo.findByIdAndDeletedAtIsNull(id)
        ?: throw NotFoundException("Lesson not found with id: $id")

}